Add Dash Apps Preview + PDF Generation

tarzzz opened the pull request:
**WIP**

tarzzz commented on src/component/plotly-dash-preview/parse.js:
```diff
@@ -0,0 +1,39 @@
+  } else {
+    // Default
+    result.format = 'png'
+  }
```
We don't need to support other image types, since these are only going to be used for PDF exports and Organize thumbnails ..

tarzzz commented on src/component/plotly-dash-preview/render.js:
```diff
@@ -0,0 +1,57 @@
+    win = null
+  })
+
+  contents.on('page-favicon-updated', () => {
```
This is a placeholder for now .. !! It will be replaced with a custom event fired by Dash App once the rendering has finished.

tarzzz commented on src/component/plotly-dash-preview/render.js:
```diff
@@ -0,0 +1,57 @@
+
+  contents.on('page-favicon-updated', () => {
+    if (PRINT_TO_PDF) {
+      contents.printToPDF({}, (err, pdfData) => {
```
TODO Error handling.. !!

tarzzz replied:
Also, the PDF Options is set to `{}` until we have some more info from: https://github.com/plotly/streambed/issues/10698#issuecomment-373161617 ..

chriddyp replied:
Could we add these options to the API? https://github.com/electron/electron/blob/master/docs/api/web-contents.md#contentsprinttopdfoptions-callback

tarzzz commented:
@etpinard Can I please get an initial review on this.. I'll be adding some tests in the meantime.. !!

tarzzz commented on test/unit/plotly-dash-preview_test.js:
```diff
@@ -0,0 +1,120 @@
+      url: 'dummy',
+      format: 'pdf'
+    }, {}, (errorCode, result) => {
+      t.ok(win.webContents.printToPDF.calledOnce)
```
This fails because `printToPDF` waits for `page-favicon-updated` event before being called. Is there a way to handle this? The next tests also fail for similar reason.. !!

etpinard replied:
`win` is an event emitter as created [here](https://github.com/plotly/image-exporter/blob/934999da692e2cdc61fdf7debdb87e022330f640/test/common.js#L25-L43), so calling `win.emit('page-favicon-updated')` should allow you to test the callback.

chriddyp commented on src/component/plotly-dash-preview/render.js:
```diff
@@ -0,0 +1,62 @@
+      })
+    } else {
+      contents.capturePage(img => {
+        result.imgData = img.toPNG()
```
interesting, i'm curious what a PNG looks like! this could be very cool for thumbnails 😸

tarzzz replied:
Yes, that's the plan. I did test it with the vanguard app:

![a](https://user-images.githubusercontent.com/1546532/37477086-cde8e7f8-289c-11e8-8349-26661e0f4168.png)

It only captures the visible area in the browser, hence the need for `width`/`height` options ..

chriddyp commented on src/component/plotly-dash-preview/parse.js:
```diff
@@ -0,0 +1,39 @@
+  }
+
+  result.width = body.width || 800
+  result.height = body.height || 600
```
This is interesting, I didn't think about this. I'm assuming that with PDF reports, where the content has absolute pixel sizing (e.g. the pages are 8.5" wide and 11" long), it doesn't matter what width and height we provide. However, for responsive apps, we'll need to provide something reasonable. Let's try a medium sized laptop: `width: 1680px`, `height: 1050`

tarzzz replied:
These options are only meant for `png` generation, as the printToPDF outputs are determined by: https://github.com/electron/electron/blob/master/docs/api/web-contents.md#contentsprinttopdfoptions-callback .. and not be the browser-size.

Although, we _can_ add them in printToPDF options after some conversion as `printToPDF` expects height/width in microns.

chriddyp replied:
Ah I understand now. OK, lets just use the electron options directly instead of doing a pixel to micron conversion.

etpinard commented:
Looking good so far. Glad someone else is trying to add things to image-expoter. That was the goal all along.

Now, by preview you mean combining reports and thumbnail generation in one component? Electron's capturePage and printToPDF work very differently and support different options. So, I'd vote for splitting this thing into two separate components.

tarzzz commented:
> Now, by preview you mean combining reports and thumbnail generation in one component?

Yes, that was the my initial plan.. 

>  So, I'd vote for splitting this thing into two separate components.

In that case, I will remove the `capturePage` for now (if it seems ok to you), since we only need PDF generation for now (right @chriddyp ? ). I can add thumbnails after `2.4.0` release.

chriddyp commented:
>  > So, I'd vote for splitting this thing into two separate components.

> In that case, I will remove the capturePage for now (if it seems ok to you), since we only need PDF generation for now

yeah, we only need PDF generation for now

tarzzz commented:
@etpinard @chriddyp: Some comments/updates:
  - we are using 'waitfor' div in dash-apps to determine if the app has finished rendering (before taking snapshot)
  - printToPDF's pdfOptions will be present in the request body and supplied to `printToPDF`.

@etpinard Please review.. 
The node (V8) tests seem to be failing for reasons unrelated to this PR.. would be great if you could take a look at them.. !!

etpinard commented on src/component/plotly-dash-preview/convert.js:
```diff
@@ -0,0 +1,22 @@
+  const result = {}
+
+  result.head = {}
+  result.head['Content-Type'] = cst.contentFormat['pdf']
```
:cow2: `cst.contentFormat.pdf` would suffice.

etpinard commented on src/component/plotly-dash-preview/index.js:
```diff
@@ -0,0 +1,10 @@
+module.exports = {
+  name: 'plotly-dash-preview',
+  ping: require('../../util/generic-ping'),
+  inject: plotlyGraph.inject,
```
No need for an `inject` method here as the `render` step has no external deps.

etpinard commented on src/component/plotly-dash-preview/parse.js:
```diff
@@ -0,0 +1,30 @@
+    sendToRenderer(code, result)
+  }
+
+  if (isNonEmptyString(body.url)) {
```
We could probably be more strict here. Using `isUrl` (like done [here](https://github.com/plotly/image-exporter/blob/a7e67453807aa08c3b27dbdfb0e81934eab5c134/src/component/plotly-graph/inject.js#L57)) would be best.

etpinard commented on src/component/plotly-dash-preview/parse.js:
```diff
@@ -0,0 +1,30 @@
+    return errorOut(400, 'invalid url')
+  }
+
+  result.pdfOptions = body.pdf_options || {}
```
What does Electron do when we pass invalid pdf options to `printToPDF`? 

- Does it ignore bad values and use the defaults?
- Does it error out?

if the latter, we should try to intercept bad pdf options here, so that we return a 400-series parse error instead of 500-series render error.

etpinard replied:
... if the former, then we don't need to do anything here :tada:

tarzzz replied:
> Does it ignore bad values and use the defaults?

That's correct. No 500s ..

etpinard replied:
Ok that's great. We won't have to worry about the various `printToPDF` options :ok_hand:

scjody replied:
It would be less frustrating to the user if we could return a 400-series error (and have that bubble up through streambed). For example if a user is trying to set the page size they may try `'page_size': 'legal'`, get back a letter option (because that's the default), and keep trying variations of `legal`. If we instead sent a 400 error saying that `page_size` is an invalid option, that will be clearer to the user.

If this is a lot of work, I don't feel strongly about it though.

etpinard replied:
I agree. Returning a 400-series error here would be best from the user's standpoint. But since we would have to duplicate how Electron validates input here in the parse step in order to get this right, this may turn into a developer's nightmare for little gain. I'd vote for leaving this as is.

scjody replied:
OK, if it's easy let's do it but I agree that this isn't worth it if it's painful or a lot of work.

etpinard commented on bin/plotly-export-server_electron.js:
```diff
@@ -53,6 +53,11 @@ const opts = {
+    {
+      name: 'plotly-dash-preview',
+      route: '/dash-preview',
+      options: plotlyJsOpts
```
No need to pass plotly.js options here. You can :hocho: this line.

etpinard commented on src/component/plotly-dash-preview/render.js:
```diff
@@ -0,0 +1,65 @@
+function render (info, opts, sendToMain) {
+  const result = {}
+
+  let win = remote.createBrowserWindow()
```
No need to specify the window width/height? That is, you're assuming all dash reports will have the same dimensions?

chriddyp replied:
Some discussion on this earlier here: https://github.com/plotly/image-exporter/pull/60#discussion_r174827073

Now that I think about it, I wonder if we should still fix the size of the window before specifying the size of the PDF export. The PDF Export size will likely change the viewport (so responsive CSS will be reflected in the PDF) but I doubt that it will re-renderer all of the javascript. So, if there are certain components that are responsive with JS and not CSS, then I doubt that they will be reflected at the new size.

So, I think it makes sense to set size of this window to be the same size as requested in the PDF renderer. For printed content, 1px will be 1/96th of an inch (https://www.w3.org/Style/Examples/007/units.en.html). So 8.5 x 11 is 816px x 1056

tarzzz replied:
Elaborating on ⬆️ my point was that:

> A4 paper measures in at 210mm by 297mm (8.26" by 11.69") whereas US Letter paper measures at 8.5" by 11" (215.9mm by 279.4mm)

Since difference is not too much between A4 and Letter, it shouldn't really affect the viewports if we change it by the `paperSize` (though we do get a significant difference when we consider A3/A2 as well)

If we consider A2, it comes out to ~`1584.0px x 2246px` .. which is big compared to general screen-sizes. (or `xvfb` screen sizes in case of headless server)

etpinard commented on src/component/plotly-dash-preview/render.js:
```diff
@@ -0,0 +1,65 @@
+          }
+        })
+      }
+    })
```
What if `finishedLoaded` doesn't resolve?

etpinard commented on src/component/plotly-dash-preview/render.js:
```diff
@@ -0,0 +1,65 @@
+    win = null
+  })
+
+  let intervalId = setInterval(() => {
```
I'm curious. Why did you decide to place the `setInterval` in the main window scope instead of in the child window scope?

Would something like:

```js
const loaded = () => {
  return contents.executeJavaScript(`
     new Promise((resolve, reject) => {
       let tries = 10
       let interval = setInterval(() => {
         const el = document.getElementById('waitfor')
         if (el) {
           clearInterval(interval)
           resolve(true)
         }
         if (--tries === 0) reject('fail to load')
       }, 500)
     })
  `) 
}

loaded().then(() => {
   contents.printToPDF(info.pdfOptios, (err, pdfData) => {
      if (err) {
        done(525)
      } else {
        result.imgData = pdfData
        done()
      }
   }).catch(() => {
     done(525)
   })
})
```

etpinard commented on test/unit/plotly-dash-preview_test.js:
```diff
@@ -0,0 +1,72 @@
+  t.test('should handle printToPDF errors', t => {
+    const win = createMockWindow()
+    sinon.stub(remote, 'createBrowserWindow').returns(win)
+    win.webContents.executeJavaScript.resolves(true)
```
Nice tests! Like mentioned in an earlier comment, we'll need to handle (and test :smirk: ) the case where `executeJavascript` does not resolve.

chriddyp commented on src/component/plotly-dash-preview/render.js:
```diff
@@ -0,0 +1,65 @@
+  const finishedLoading = () => {
+    return win.webContents.executeJavaScript(`
+      new Promise((resolve, reject) => {
+        var a = document.getElementById("waitfor");
```
Can we make this a variable? i.e. part of the API where the use can pass in an arbitrary CSS selector. That would modify this to be `document.querySelector`. 

For example, if the app has a graph in it, I might end up using a query selector like `#mygraph .plot-container` in order to wait until the graph was rendered (as plotly.js draws `.plot-container` dynamically). But, if the report doesn't have a graph in it, then I'd inject my own div with id `waitfor`

tarzzz replied:
Just to confirm, the query selector comes as the part of the incoming POST request (with `.plot-container` as default if nothing is provided)? 

It should be safe to assume the dash-app has atleast one chart even if the incoming request does not specify it.

tarzzz commented:
@etpinard @chriddyp Updates made:

  - Atleast one of `loading_selector` or `timeout` should be specified in request. If `timeout` is specified, we wait for this time before capturing the page. 
  - `pageSize` should be specified either as the part of the request `body` or as the part of `pdfOptions`. We use it to determine the browser window size as indicated by Chris in https://github.com/plotly/image-exporter/pull/60#discussion_r177200389
  - Added tests for these ⬆️ options as well as to check for rejected promises (`executeJavascript`.
  - Addressed minor code-style related comments.

etpinard commented on src/component/plotly-dash-preview/parse.js:
```diff
@@ -24,6 +24,13 @@ function parse (body, opts, sendToRenderer) {
   }
 
   result.pdfOptions = body.pdf_options || {}
+  if (!body.loading_selector && !body.timeout) {
```
We should use utils `isPositiveNumber` for `timeout` and `isNonEmptyString` for `loading_selector`.

Oh and I would vote for just `selector` instead of `loading_selector` which is unnecessarily verbose here I think.

etpinard commented on src/component/plotly-dash-preview/render.js:
```diff
@@ -31,35 +33,53 @@ function render (info, opts, sendToMain) {
+        let tries = ${cst.maxRenderingTries}
+
+        if (${info.timeOut}) {
+          tries = parseInt(${info.timeOut} * 1000/${cst.minInterval})
```
Let's move the coercion step to `parse`. Note that `Number(info.timeOut)` would suffice. `parseInt` is slow.

etpinard replied:
:cow2: `1000 / ${cst.minInterval)` please.

etpinard commented on src/component/plotly-dash-preview/render.js:
```diff
@@ -31,35 +33,53 @@ function render (info, opts, sendToMain) {
         }
+
+        let interval = setInterval(() => {
+          let el = document.querySelector("${info.loadingSelector}")
```
:cow2: use `'` over `"` even inside \`\`

etpinard commented on src/component/plotly-dash-preview/render.js:
```diff
@@ -31,35 +33,53 @@ function render (info, opts, sendToMain) {
     })
-  }, 500)
+  }).catch(() => {
+    done(525)
```
Sorry to not having noticed this before. But, it would be nice to separate `printToPDF` errors (that's the `done(525)` inside the `printToPDF` callback) and this one which is the result of a timeout. 

I propose making `525` the `printToPDF` callback error, and adding `526` when we hit a timeout.

tarzzz replied:
``` 
526: 'plotly.js version 1.11.0 or up required',
```
526 is already taken.. maybe traditional `408 Request Timeout` or `504 Gateway Timeout`?

etpinard replied:
It's already taken in the `plotly-graph` component, right? Not in `plotly-dash-preview`? I don't like overriding 504 (a gateway timeout is something else) and I want this to be a 500-series.

etpinard commented on package.json:
```diff
@@ -49,7 +49,7 @@
     "glob": "^7.1.2",
     "is-plain-obj": "^1.1.0",
-    "is-url": "^1.2.2",
+    "is-url": "1.2.2",
```
Ok yeah. That might work. But really, we should be using the latest version of `is-url`. Looking at their [changelog](https://github.com/segmentio/is-url/commits/master), bumping to version 1.2.4 should fix everything.

etpinard commented on src/component/plotly-dash-preview/parse.js:
```diff
@@ -31,6 +32,19 @@ function parse (body, opts, sendToRenderer) {
+  } else {
+    return errorOut(400, 'pageSize must either be A3, A4, A5, Legal, Letter, ' +
+                         'Tabloid or an Object containing height and width ' +
+                         'in microns.')
```
:cow2: I really see indents like this in JS. That seems very pythonic. I would write this as:

```js
errorOut(
  400,
  'pageSize must either be A3, A4, A5, Legal, Letter, ' +
  'Tabloid or an Object containing height and width ' +
  'in microns.'
)
```

etpinard commented on src/component/plotly-dash-preview/parse.js:
```diff
@@ -31,6 +32,19 @@ function parse (body, opts, sendToRenderer) {
 
+  if (cst.sizeMapping[result.pdfOptions.pageSize]) {
+    result.browserSize = cst.sizeMapping[result.pdfOptions.pageSize]
+  } else if (body.pageSize && body.pageSize.width && body.pageSize.height) {
```
Use `isPositiveNumeric(body.pageSize.width)` and `isPositiveNumeric(body.pageSize.height)` to error early on 0 and negative inputs.

etpinard commented on test/unit/plotly-dash-preview_test.js:
```diff
@@ -23,6 +24,28 @@ tap.test('parse:', t => {
+  })
+
+  t.test('should error when pageSize is not given', t => {
+    
```
:zap:

etpinard commented on test/unit/plotly-dash-preview_test.js:
```diff
@@ -23,6 +24,28 @@ tap.test('parse:', t => {
   })
 
+  t.test('should error when neither loading_selector or timeout is given', t => {
+    
```
:zap:

etpinard commented on src/component/plotly-dash-preview/parse.js:
```diff
@@ -25,24 +27,29 @@ function parse (body, opts, sendToRenderer) {
-  if (!body.loading_selector && !body.timeout) {
-    return errorOut(400, 'either loading_selector or timeout must be specified')
+  if (!isNonEmptyString(body.selector) && !isPositiveNumeric(body.timeout)) {
+    return errorOut(400, 'either selector or timeout must be specified')
```
This here may be overly stringent. Perhaps we could default to a 3000ms timeout if `timeout` and `selector` are omitted from the POST request. @chriddyp what do you think?

etpinard commented on src/component/plotly-dash-preview/render.js:
```diff
@@ -35,14 +35,10 @@ function render (info, opts, sendToMain) {
-        if (${info.timeOut}) {
-          tries = parseInt(${info.timeOut} * 1000/${cst.minInterval})
-        }
+        let tries = info.tries || ${cst.maxRenderingTries}
```
Very clean here. Nice :ok_hand:

tarzzz commented:
Added integration test, and ran coverage:
![image](https://user-images.githubusercontent.com/1546532/38074517-962b8978-334c-11e8-9644-442eea725c4a.png)

tarzzz commented:
Updates: Added integration test, and `timeout` code (526) for rendering related timeouts.

Still blocked by intermittently failing integration tests..

etpinard commented on src/component/plotly-dash-preview/render.js:
```diff
@@ -23,7 +23,7 @@ function render (info, opts, sendToMain) {
 
     if (errorCode) {
-      result.msg = 'dash preview generation failed'
+      result.msg = cst.statusMsg[errorCode]
```
nicely done :+1:

etpinard commented:
@tarzzz thanks for running `npm run coverage`. You can ignore the uncovered lines in `plotly-dash-preview/convert.js`. But it would be nice to bring up `plotly-dash-preview/parse.js` to 100% coverage.

As for our plotly-export-server integration tests. Perhaps we should skip them on CI for now (and open a new issue about it). I think booting up 6 browser windows on the CI machines might be a little too much for them to handle. We could revisit this later.

tarzzz commented:
@etpinard : Coverage ..

![image](https://user-images.githubusercontent.com/1546532/38096556-c478ff7e-3390-11e8-83a1-fbcfa7cd0d6c.png)

Added issue to address intermittent errors: https://github.com/plotly/image-exporter/issues/61
